### ISSUE TITLE
[EPIC] Phase out MM and use V10 DataModels instead

### ISSUE THREAD

**Issue by whitespine** [Eranziel/foundryvtt-lancer#429]
FRIENDSHIP ENDED WITH MACHINE MIND. NOW FOUNDRY NATIVE IS MY BEST FRIEND.

I've brought this up on this discord, and I think whether we actually do this is not a foregone conclusion. But I want to make it as clear as possible what would be involved.

To that end,  I figured I'd start enumerating what I would envision the steps to be, from where we currently are to a fully viable V10 edition. I'm

- [] Create data model schemas for our template.json
  - [] Replace all refs with UUID strings
  - [] Should make reusable submodels for:
    - [x] Counter
    - [x] Action 
    - [x] Bonus 
    - [x] Synergy 
    - [x] Range
    - [x] Damage 
    - [] DamageTypeChecklist - Does not yet exist as a custom field type
    - [] RangeTypeChecklist - ditto
    - [] All other checklists - ditto
  - [x] Make classes to implement these subclasses
  - ~Most templates can remain as is~ Most templates were streamlined, however we will need to add new derived and accessory properties for active effects enabled bonus computation, particularly
    - [] Boolean Flags for all statuses / conditions
    - [] Repair cost
    - [] Item aggregations
    - ...?
- [] Create separate classes for every item type
  - [] Should probably have a common base class / mixing
  - Todo: put a more specific Todo list here
- [] Update paths everywhere. For ease of distribution of work, we separate this into
  - [] Handlebars template data paths
  - [] Handlebars template action paths
- [] Update handlebars logic to not use writeback 
  - [] .update() handling/triggering logic
    - [] Isolate target to furthest along path
    - [] Truncate paths to be relative to that object
  - [x] Update all ref embedding to use uuids instead
  - [] Fix gen actions
- [] Re-evaluate if custom drag drop logic is still necessary with new drag drop workflows
- [x] Unify "native" drop and "ref" drop logic under new foundry entity drag drop logic
- [] Make logic for converting Bonuses to equivalent active effects
  - [] "Override" should be implemented as just a higher priority "set" effect
  - Weapon effects that are specific to types and ranges will be a bit trickier. 
    - Should these remain as lists of more complex objects?
    - Should they be broken down into 25 pairings akin to "bonus_range_blast_energy"?
    - Both kinda suck 
- [] Make logic for propagating ActiveEffects from pilots to mechs to deployables
  - [] Ensure this logic is triggered/cascaded on update
  - [] Guard against circular loops
- [x] Remake prepared at to be less awful (do this later - we don't know our requirements)
  - [x] Work has begun on this
- [] Find a solution to displaying asynchronously retrieved information inline
  - [] check if Svelte stuff we have would work here?
  - [] async loaded tooltips
- [] Make our "core" conditions apply simple active effects to set their corresponding statuses and conditions
- [] Implement logic to convert our unpacked MM objects to our new foundry items/actors/journals.
- [] Replace regrefs with a new UUID type everywhere.
  - Keep a "type" so we can at least prepare scaffolding to display things and have typed "slots"
- [x] Investigate if there are solutions for generating type definitions based on DataModel definitions -- It isn't, I did them all manually though

**Comment by mason-bially** [comment]
So a side not of something that came up is that lancer's item model seems a little different than the other systems. Specifically [Maestro](https://github.com/death-save/maestro) and [Automated Animations](https://github.com/otigon/automated-jb2a-animations) among others. Notably these have two issues:

- Currently we don't put the item id into chat cards properly. I hope to fix that with my Chat Card UI refactor.
- The way we currently do items means that **every single copy of the item** must be modified to bind effects to it. Which doesn't appear to be the intent of how these modules should work, or how it works in other systems. This implies to me that we are doing something fundamentally different with our current data model. That's an issue that this ticket should look into hopefully fixing as well.

**Comment by BoltsJ** [comment]
Re: Create separate classes for every item type

I've done some experimenting with this for actors. Once the v10 branch is merged into `dev` I can star making the boilerplate for this.

**Comment by CoalPoweredPuppet** [comment]
Will this help get Lancer up to V10?  If so, how can I help.  I only have self-taught html and css knowledge, but I will do what I can.

**Comment by dodgepong** [comment]
Not directly. v10 support can come without this, though we will definitely want to do this eventually.

**Comment by CoalPoweredPuppet** [comment]
This seems like the wrong place to ask this, but where do I go to try to help the v10 support?  And how can I help you do...whatever this is?

**Comment by Eranziel** [comment]
I think we can consider this closed. (Hooray!) The dev branch with data models is working well overall, bughunting is being tracked in #526.
